API/ev3_robot.c

```diff
@@ -0,0 +1,163 @@
+
+}
+
+void RobotInit(int ColorLeft, int ColorRight, uint8_t MotorLeft, uint8_t MotorRight, int Gyro, bool Debug){
+	
+	PoseInit();
+	SetKPID(0,0,0);
+
+	MyRobot.WheelCircunference = MyRobot.WheelDiameter * M_PI;
+	MyRobot.ColorLeft = ColorLeft;
+	MyRobot.ColorRight = ColorRight;
+	MyRobot.MotorLeft = MotorLeft;
+	MyRobot.MotorRight = MotorRight;
+	MyRobot.MotorDual = MotorLeft + MotorRight;
+	MyRobot.Gyro = Gyro;
+	MyRobot.Kfriction = 1.0;
+	
+	int reflectedLightLeft = ReadEV3ColorSensorReflectedLight(MyRobot.ColorLeft);
+	int reflectedLightRight = ReadEV3ColorSensorReflectedLight(MyRobot.ColorRight);
+	
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroHardwareCommand0x88); // chack Gyro Veersion //
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroHardwareReboot); // 2.5s, valid for all versions
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroSoftwareOffset); /* chack for proper reset */
+	wait(100);
+	int angle = ReadEV3GyroSensorAngle(MyRobot.Gyro, EV3GyroNormalAngle);	
+
+	ResetRotationCount(MyRobot.MotorDual);
+
+	int rotationsLeft = MotorRotationCount(MyRobot.MotorLeft);
+	int rotationsRight = MotorRotationCount(MyRobot.MotorRight);
+
+	if (Debug){
+ 		LcdClean();
+        LcdTextf(1, 0, LcdRowToY(1), "Color Left: %d", reflectedLightLeft);
+		LcdTextf(1, 0, LcdRowToY(2), "Color Right: %d", reflectedLightRight);
+        LcdTextf(1, 0, LcdRowToY(3), "Angle: %d", angle);
+        LcdTextf(1, 0, LcdRowToY(4), "Rotations Left: %d", rotationsLeft);
+        LcdTextf(1, 0, LcdRowToY(5), "Rotations Right: %d", rotationsRight);
+        Wait(100);
+	}
+	
+}
+
+int Travelcm(int distance){
+
+	return (int)(distance * MyRobot.Kfriction / MyRobot.WheelCircunference);
+
+}
+
+int StraightGyroDegrees(int distDegre, int angle, int speed, bool brake){
+
+int traveled = 0;
+int rotationsLeft = MotorRotationCount(MyRobot.MotorLeft);
+int rotationsRight = MotorRotationCount(MyRobot.MotorRight);
+int angleNow = ReadEV3GyroSensorAngle(MyRobot.Gyro, EV3GyroNormalAngle);
+int error;
+int errorD = 0; // error compensation for derivative part
+int oldErrorD= 0;
+int errorI = 0; //accumalated errors for integral part
+
+int uOut; // control for motors
+
+OnFwdSync(MyRobot.MotorDual, speed);
+do{
+	//desviation measurement
+	angleNow = ReadEV3GyroSensorAngle(MyRobot.Gyro, EV3GyroNormalAngle);
+	
+	error = angle - angleNow; // For Proportional part
+	errorD = error - oldErrorD; // For Darivative part
+	errorI = errorI + error; //For Integral part
+	//PID control
+	uOut = (int)((error *  StraightPid.Kp) + (errorI * StraightPid.Ki) + (errorD * StraightPid.Kd ));
+	oldErrorD = error;
+	
+	/* 
+	 * Proportional part
+	 * (error ^  StraightPid.Kp)
+	 * Integral part
+	 * (errorI * StraightPid.Ki
+	 * Derivative part
+	 * (errorD * StraightPid.Kd )
+	 *  Note; Delta t is considered in Ki, Kd values Ki*Delta t and Kd/Delta t
+	 */
+
+	// new power to motors
+	// need to cheek max/min speed?
+	OnFwdReg(MyRobot.MotorLeft, speed + uOut);
+	OnFwdReg(MyRobot.MotorRight, speed - uOut);
```

Comment:
Our strategy at EV3-G was to increase speed by 45 degrees and decrease in similar way, this was precise, but it is slow, every turn, every change of direction, the robot stops before.
We will try to code  our EV3-G block.
Many, many thanks for the suggestion of function RotateMotorRampNoWait()
we are still looking for documentation about ev3_output.

Reply:
If you can avoid braking altogether (indeed, the current version allows this), that would be the fastest. For the stopping before the target, it may be necessary to have some minimum continuous speed level at the end of the deceleration ramp. It's true that it wouldn't speed it up though.

Re documentation: it's quite low-level, but the kernel stuff that `ev3_output.c` in the end calls is here: https://github.com/mindboards/ev3sources-xtended/blob/b32a23625be02eb22f23ac45d2ef3bd4a2a9173f/ev3sources/lms2012/d_pwm/Linuxmod_AM1808/d_pwm.c#L2146 There should be some similarities to EV3-G blocks because they have to go through the kernel too.

Reply:
Thanks a buch!!!
A last question about output.c, in MS VSCODE show me the error:
{
	"resource": "/c:/GitHub/EV3-API/API/ev3_output.c",
	"owner": "C/C++",
	"severity": 8,
	"message": "the identifier \"O_SYNC\" is not defined",
	"startLineNumber": 140,
	"startColumn": 66,
	"endLineNumber": 140,
	"endColumn": 72
}

Reply:
I think you can remove the problematic `O_SYNC`, it should have no effect there. I don't know why the warning appeared, but it might be that the include path set used inside vscode isn't the same as the one the compiler uses.